posts/fall2019/js-chats-3/index.md

```diff
@@ -0,0 +1,1285 @@
+---
+date: "2019-10-22"
+title: "A Gentle Introduction to React Hooks"
+subtitle: "JavaScript Chats with ACM Hack Session 3"
+---
+
+- [Quick Review of the `class` Lifecycle Methods](#quick-review-of-the-class-lifecycle-methods)
+- [Quick Review of Function Component](#quick-review-of-function-component)
+- [Overview of Hooks](#overview-of-hooks)
+  - [`useState`: State in Function Components](#usestate-state-in-function-components)
+  - [🍏Application: "Hey React, Set an Alarm" with Multiple `useState`](#application-hey-react-set-an-alarm-with-multiple-usestate)
+  - [Introduction to Custom Hooks](#introduction-to-custom-hooks)
+  - [`useEffect`: Side Effect in Function Component](#useeffect-side-effect-in-function-component)
+  - [🍏Application: "Applying to Boogle" with `useEffect`](#application-applying-to-boogle-with-useeffect)
+- [2 Rules of Hooks](#2-rules-of-hooks)
+  - [A List Model of Keeping States](#a-list-model-of-keeping-states)
+- [🍏Application: "Power of Hooks" with Debouncing](#application-power-of-hooks-with-debouncing)
+  - [`setTimeout` and `clearTimeout`](#settimeout-and-cleartimeout)
+  - [Implementation](#implementation)
+  - [Cleaning up `useEffect`](#cleaning-up-useeffect)
+  - [Abstract, Reuse, Custom Hook](#abstract-reuse-custom-hook)
+  - [What's Wrong with the Old `class` Way?](#whats-wrong-with-the-old-class-way)
+- [Final Challenge](#final-challenge)
+
+## Quick Review of the `class` Lifecycle Methods
+
+Before we dive into hooks, let's review some lifecycle methods
+in the old `class` syntax. 
+Here are some important ones that we need to know: 
+
+- `setState`: update the state within a component, triggers `render` 
+- `componentDidMount`: called when a component is just mounted.
+  `componentDidMount` is allowed to perform side-effect here such as API
+  calls, and state updates with `setState`
+- `render`: React calls `render` to get the updated version of the component
+- `componentDidUpdate`: called after the component re-rendered. Side effects allowed
+
+The lifecycle methods are well summarized in the following flow chart:
+
+![react lifecycle](images/lifecycle.png)
+
+> See the diagram at: http://projects.wojtekmaj.pl/react-lifecycle-methods-diagram/
+
+Exercise: Build a simple counter component, but the initial value 
+of the counter is given by the server. Here we emulate an API call
+to fetch the initial value.
+
+```jsx
+// faking a call to the server to get an initial value
+function getInitialCounterValue() {
+  return new Promise(resolve => setTimeout(() => resolve(3)), 200);
+}
+```
+
+<details>
+
+<summary> Click to see solution </summary>
+
+  ```jsx
+  class Counter extends React.Component {
+    state = { count: 0 }
+
+    constructor() {
+      this.onIncrement = this.onIncrement.bind(this);
+    }
+
+    componentDidMount() {
+      getInitialCounterValue().then(count => {
```

Comment:
Use an async function here?

Also grep for `.then` in the file and change the rest to use async functions.

Reply:
I intentionally kept myself away from async since later I asked the question what if we use async in `useEffect`.